api.lisp

```diff
@@ -1,130 +1,217 @@
-            )
-        )
-    )
-)
-
-(defun find-object (list name)
-    (if list
-        (let ((obj (car list)))
-            (if (string= name (gethash "name" obj)) obj (find-object (cdr list) name)))
-        nil))
-
-(defun find-object-in-config-list (config list-key name)
-    (let ((objects (gethash list-key config)))
-        (find-object objects name)))
-
-(defun get-context (config context)
-    (find-object-in-config-list config "contexts" context))
-
-(defun get-user (config user)
-    (find-object-in-config-list config "users" user))
-
-(defun get-cluster (config cluster)
-    (find-object-in-config-list config "clusters" cluster))
-
-(defun current-context (config)
-    (let ((context (get-context config (gethash "current-context" config))))
-        (if (not (eq nil context))
-            (gethash "context" context)
-            nil)))
-
-(defun current-user (config)
-    (let ((context (current-context config)))
-       (if context
-           (gethash "user" (get-user config (gethash "user" context)))
-           nil)))
+  "Load the default configuration in this environment.
+
+First, try to load configuration from one or more files listed in the
+KUBECONFIG environment variable, if that variable is set.
+
+Otherwise, try to read the configuration file from the home user
+directory (~/.kube/config).
+
+If this file does not exist, return a default configuration."
+  (let ((source (or (split (inter-directory-separator)
+                           (getenvp "KUBECONFIG"))
+                    (probe-file (merge-pathnames #P".kube/config"
+                                                 (user-homedir-pathname))))))
+    (if source
+        (load-config source)
+        (alist-hash-table
+         '(("apiVersion" "v1")
+           ("kind" "Config"))
+         :test #'equal))))
+
+(defmacro define-accessors (name &optional (key (string-downcase name)))
+  "Define both NAME and (SETF NAME) to access slot KEY in objects.
+
+KEY is a string that defaults to the NAME symbol (downcased), and
+represents the name of the key to access in an object to retrieve its
+value."
+  (check-type name symbol)
+  (check-type key string)
+  (let ((object (copy-symbol :object))
+        (value (copy-symbol :value)))
+    `(progn
+       ;; Inline accessors: they are not going to change at runtime.
+       (declaim (inline ,name (setf ,name)))
+
+       ;; Reader: allow NIL object.
+       (defun ,name (,object)
+         ,(format
+           nil
+           "Get the value associated with ~S in OBJECT, or NIL. ~%~%~
+
+            If OBJECT is NIL, returns NIL."
+           key)
+         (and ,object (gethash ,key ,object)))
+
+       ;; Writer .
+       (defun (setf ,name) (,value ,object)
+         ,(format
+           nil
+           "Set the value associated with ~S in OBJECT to VALUE."
+           key)
+         (setf (gethash ,key ,object) ,value)))))
+
+;; values
+(define-accessors name)
+(define-accessors server)
```

Comment:
You are missing:

```
(define-accessors context)
(define-accessors cluster)
(define-accessors current-user)
```

and possibly others?

Reply:
Indeed, sorry. 

By the way, I think the distinction between `context-name` and `context` is not as useful as I first thought, and can even be confusing. I'll update the code to use the same keys as in the data format.

I couldn't test the examples locally. As far as I know, the recommended way would be to install minikube. Is that right?